"incompatible type" when mixing float32 and cfloat in generics

Mixing `float32` and `cfloat` can sometimes lead to a strange error that exposes the generated C code.

### Example
```nim
type
  Vec3[T: SomeNumber] = object
    arr: array[3, T]

var 
  cfloatArr: array[3, array[3, cfloat]]
  cfloatSeq = newSeq[Vec3[cfloat]]()
for row in cfloatArr:
  cfloatSeq.add(Vec3[float32](arr: [row[0], row[1], row[2]]))
echo cfloatSeq
```

### Current Output
```
/Users/sekao/.cache/nim/nim_misc_d/@mnim_misc.nim.c:449:40: error: assigning to 'tyObject_Vec3__a2hI8d3noWjXtSOVv1UfRA' (aka 'struct tyObject_Vec3__a2hI8d3noWjXtSOVv1UfRA') from
      incompatible type 'tyObject_Vec3__c79bdlxFEYQ7KlCcTS4NeNw' (aka 'struct tyObject_Vec3__c79bdlxFEYQ7KlCcTS4NeNw')
                                cfloatSeq__nim95misc_37->data[T5_] = T4_;
```

### Expected Output
```
@[(arr: [0.0, 0.0, 0.0]), (arr: [0.0, 0.0, 0.0]), (arr: [0.0, 0.0, 0.0])]
```

### Additional Information
The code works in nim 1.2.6.

```
$ nim -v
Nim Compiler Version 1.6.2 [MacOSX: amd64]
Compiled at 2021-12-17
Copyright (c) 2006-2021 by Andreas Rumpf
```

--- a/compiler/types.nim
+++ b/compiler/types.nim
@@ -1149,13 +1149,14 @@ proc sameTypeAux(x, y: PType, c: var TSameTypeClosure): bool =
   of tyEmpty, tyChar, tyBool, tyNil, tyPointer, tyString, tyCstring,
      tyInt..tyUInt64, tyTyped, tyUntyped, tyVoid:
     result = sameFlags(a, b)
-    if result and PickyCAliases in c.flags:
+    if result and {PickyCAliases, ExactTypeDescValues} <= c.flags:
       # additional requirement for the caching of generics for importc'ed types:
       # the symbols must be identical too:
       let symFlagsA = if a.sym != nil: a.sym.flags else: {}
       let symFlagsB = if b.sym != nil: b.sym.flags else: {}
       if (symFlagsA+symFlagsB) * {sfImportc, sfExportc} != {}:
         result = symFlagsA == symFlagsB
+
   of tyStatic, tyFromExpr:
     result = exprStructuralEquivalent(a.n, b.n) and sameFlags(a, b)
     if result and a.len == b.len and a.len == 1:

--- a/tests/alias/t19349.nim
+++ b/tests/alias/t19349.nim
@@ -1,0 +1,19 @@
+discard """
+  action: "compile"
+"""
+
+type
+  Vec3[T: SomeNumber] = object
+    arr: array[3, T]
+
+var 
+  cfloatArr: array[3, array[3, cfloat]]
+  cfloatSeq = newSeq[Vec3[cfloat]]()
+for row in cfloatArr:
+  cfloatSeq.add(Vec3[float32](arr: [row[0], row[1], row[2]]))
+
+var 
+  cuintArr: array[3, array[3, cuint]]
+  cuintSeq = newSeq[Vec3[cuint]]()
+for row in cuintArr:
+  cuintSeq.add(Vec3[uint32](arr: [row[0], row[1], row[2]]))

